orca-webhook/src/main/java/com/netflix/spinnaker/orca/webhook/tasks/MonitorWebhookTask.java

```diff
@@ -119,135 +126,143 @@ public TaskResult execute(StageExecution stage) {
-      } catch (PathNotFoundException e) {
-        monitor.put(
-            "error",
-            String.format(JSON_PATH_NOT_FOUND_ERR_FMT, "progress", stageData.statusJsonPath));
-        return TaskResult.builder(ExecutionStatus.TERMINAL).context(responsePayload).build();
-      }
-      if (!(progress instanceof String)) {
-        monitor.put(
-            "error",
-            String.format(
-                "The json path '%s' did not resolve to a String value",
-                stageData.progressJsonPath));
-        monitor.put("resolvedValue", progress);
-        return TaskResult.builder(ExecutionStatus.TERMINAL).context(responsePayload).build();
-      }
-      if (progress != null) {
-        monitor.put("progressMessage", progress);
+                  "The json path '%s' did not resolve to a String value",
+                  stageData.progressJsonPath));
+          monitor.setResolvedValue(progress);
+          return TaskResult.builder(ExecutionStatus.TERMINAL).context(context).build();
+        } else {
+          monitor.setProgressMessage((String) progress);
+        }
       }
-    }
 
-    var statusMap =
-        createStatusMap(
-            stageData.successStatuses, stageData.canceledStatuses, stageData.terminalStatuses);
-
-    if (result instanceof Number) {
-      var status = result.equals(100) ? ExecutionStatus.SUCCEEDED : ExecutionStatus.RUNNING;
-      monitor.put("percentComplete", result);
-      return TaskResult.builder(status).context(responsePayload).build();
-    } else if (statusMap.containsKey(result.toString().toUpperCase())) {
-      return TaskResult.builder(statusMap.get(result.toString().toUpperCase()))
-          .context(responsePayload)
-          .build();
+      if (result instanceof Number) {
+        var status = result.equals(100) ? ExecutionStatus.SUCCEEDED : ExecutionStatus.RUNNING;
+        monitor.setPercentComplete((Number) result);
+        return TaskResult.builder(status).context(context).build();
+      } else {
+        var statusMap =
+            createStatusMap(
+                stageData.successStatuses, stageData.canceledStatuses, stageData.terminalStatuses);
+        ExecutionStatus status =
+            statusMap.getOrDefault(result.toString().toUpperCase(), ExecutionStatus.RUNNING);
+        return TaskResult.builder(status).context(context).build();
+      }
     }
-
-    return TaskResult.builder(ExecutionStatus.RUNNING).context(responsePayload).build();
   }
 
   @Override
   public void onCancel(@Nonnull StageExecution stage) {
     webhookService.cancelWebhook(stage);
   }
 
+  private boolean shouldRetry(HttpStatus statusCode, WebhookStage.StageData stageData) {
+    int status = statusCode.value();
+    var retries = stageData.getRetries();
+
+    if (retries != null && retries.containsKey(status)) {
+      // specific retry limit configured
+      var retryConfig = retries.get(status);
+      long pastStatuses =
```

Comment:
nit; `pastStatuses` reads like a list of statuses, maybe change to `attempts`?

Reply:
I updated this and extracted a method to make it read a little better